osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs

```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
             });
         }
 
+        private readonly object joinOrLeaveTaskLock = new object();
+        private Task? joinOrLeaveTask;
+
         /// <summary>
         /// Joins the <see cref="MultiplayerRoom"/> for a given API <see cref="Room"/>.
         /// </summary>
         /// <param name="room">The API <see cref="Room"/>.</param>
         public async Task JoinRoom(Room room)
         {
-            if (Room != null)
-                throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
+            Task? lastTask;
+            Task newTask;
 
-            Debug.Assert(room.RoomID.Value != null);
+            lock (joinOrLeaveTaskLock)
+            {
+                lastTask = joinOrLeaveTask;
+                joinOrLeaveTask = newTask = Task.Run(async () =>
+                {
+                    if (lastTask != null)
```

Comment:
Now that I think about it, there might be a theoretical data race here? What I'm thinking is that this `lastTask` read is not under locking, so it might end up reading `newTask`, or even another task entirely, although I'm not sure as I don't quite know how lambda capturing will work in this case...

Reply:
It can't, it's a local variable.